config/file_config.go

```diff
@@ -435,603 +255,494 @@ func (f *fileConfig) GetListenAddr() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.config.GetBool("PeerManagement.UseTLS"), nil
+	return f.mainConfig.PeerManagement.UseTLS, nil
 }
 
 func (f *fileConfig) GetUseTLSInsecure() (bool, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.config.GetBool("PeerManagement.UseTLSInsecure"), nil
+	return f.mainConfig.PeerManagement.UseTLSInsecure, nil
 }
 
 func (f *fileConfig) GetIdentifierInterfaceName() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.config.GetString("PeerManagement.IdentifierInterfaceName"), nil
+	return f.mainConfig.PeerManagement.IdentifierInterfaceName, nil
 }
 
 func (f *fileConfig) GetUseIPV6Identifier() (bool, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.config.GetBool("PeerManagement.UseIPV6Identifier"), nil
+	return f.mainConfig.PeerManagement.UseIPV6Identifier, nil
 }
 
 func (f *fileConfig) GetRedisIdentifier() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.config.GetString("PeerManagement.RedisIdentifier"), nil
+	return f.mainConfig.PeerManagement.RedisIdentifier, nil
 }
 
 func (f *fileConfig) GetHoneycombAPI() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.conf.HoneycombAPI, nil
+	return f.mainConfig.HoneycombAPI, nil
 }
 
 func (f *fileConfig) GetLoggingLevel() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.conf.LoggingLevel, nil
+	return f.mainConfig.LoggingLevel, nil
 }
 
 func (f *fileConfig) GetLoggerType() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.conf.Logger, nil
+	return f.mainConfig.Logger, nil
 }
 
 func (f *fileConfig) GetHoneycombLoggerConfig() (HoneycombLoggerConfig, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	hlConfig := &HoneycombLoggerConfig{}
-	if sub := f.config.Sub("HoneycombLogger"); sub != nil {
-		err := sub.UnmarshalExact(hlConfig)
-		if err != nil {
-			return *hlConfig, err
-		}
-
-		hlConfig.LoggerAPIKey = f.config.GetString("HoneycombLogger.LoggerAPIKey")
-
-		// https://github.com/spf13/viper/issues/747
-		hlConfig.LoggerSamplerEnabled = f.config.GetBool("HoneycombLogger.LoggerSamplerEnabled")
-		hlConfig.LoggerSamplerThroughput = f.config.GetInt("HoneycombLogger.LoggerSamplerThroughput")
-
-		v := validator.New()
-		err = v.Struct(hlConfig)
-		if err != nil {
-			return *hlConfig, err
-		}
-
-		return *hlConfig, nil
-	}
-	return *hlConfig, errors.New("No config found for HoneycombLogger")
+	return *f.mainConfig.HoneycombLogger, nil
 }
 
 func (f *fileConfig) GetCollectorType() (string, error) {
 	f.mux.RLock()
 	defer f.mux.RUnlock()
 
-	return f.conf.Collector, nil
+	return f.mainConfig.Collector, nil
 }
 
 func (f *fileConfig) GetAllSamplerRules() (map[string]interface{}, error) {
 	samplers := make(map[string]interface{})
 
-	keys := f.rules.AllKeys()
-	for _, key := range keys {
-		parts := strings.Split(key, ".")
+	_, err := readConfigInto(samplers, f.opts.RulesLocation, f.opts)
```

Comment:
Is it necessary to read each time or can it be read once per unique hash like the other rules/config?

Reply:
I had a justification for doing it this way (we need to read into different data types at different times). But now that you've poked me about it, I think I have a better idea. Thanks.